### ISSUE TITLE
Change schema for finding repositories on USB sticks in OstreeRepoFinderMount

### ISSUE THREAD

**Issue by pwithnall** [ostreedev/ostree#1174]
tl;dr: I want to make a backwards-incompatible change to how `OstreeRepoFinderMount` looks for repos on USB sticks. That should be fine though, since it’s experimental API.

I’ve been putting together a tool for pulling an OSTree onto a USB stick so that `OstreeRepoFinderMount` can find it (which will be a separate PR). While doing so, I started thinking about live USBs and how this will interact with those. Ideally, a live USB stick containing an OSTree-based OS could be found by `OstreeRepoFinderMount` without any additional work. The current schema for finding repositories which `OstreeRepoFinderMount` uses can’t support that — it requires a load of `$usb_stick/.ostree/repos/$collection_id/$ref_name` symlinks to `$usb_stick/ostree/repo`.

Thinking about that a bit more, I realised that the current scheme won’t pick up new refs from the live USB’s repository if (for example) someone boots from the live USB, pulls a new ref into the repository (for example, by installing a flatpak app), then plugs the USB stick into another computer and tries to find that ref on the stick using `OstreeRepoFinderMount`. The ref will exist in the live USB’s repository, but there will be no `$usb_stick/.ostree/repos/$collection_id/$ref_name` symlink for it.

I think that requiring a symlink for each collection–ref was a mistake. It was intended to allow multiple independent repositories to coexist on a single USB stick, but I think a better way of doing this is to have a `$usb_stick/.ostree/repos.d` directory containing symlinks to the repositories (or the repositories themselves), as children. The names of the directories/symlinks in `repos.d` doesn’t matter, though we could perhaps access them in lexical order for predictability.

In addition to this, I think `OstreeRepoFinderMount` should also always look for `$usb_stick/ostree/repo` as a repository, since that’s a well-known path for OSTree repositories, and sufficiently common that it seems a bit pointless to need well-known symlink to it from `$usb_stick/.ostree/repos.d`.

Two questions:
 - Is the experimental API break OK? I assume it will be.
 - Does the new schema sound reasonable?

**Comment by cgwalters** [comment]
The old scheme seemed weird to me for sure, I doubt anyone else is using it.

Before I forget, a while ago we added https://github.com/ostreedev/ostree/pull/982 - this is related in that when we start talking about local drives/repos we can be more efficient in our imports.  Now by default that inherits the "local repos are trusted" logic - I presume for your use case that shouldn't be the default, so...probably all repo finder pulls should just use `OSTREE_REPO_PULL_FLAGS_UNTRUSTED`?

Anyways...no opposition to supporting multiple repos, but maybe make the happy path nicer and also support `$usb_stick/.ostree/repo` ?

But in this USB case you're talking about, would there be some logic on the mounting computer to replicate the chain of mounts for live USB?  At least in Fedora it's quite [interesting](https://rwmj.wordpress.com/2009/07/15/unpack-the-russian-doll-of-a-f11-live-cd/).

**Comment by pwithnall** [comment]
> Before I forget, a while ago we added #982 - this is related in that when we start talking about local drives/repos we can be more efficient in our imports. Now by default that inherits the "local repos are trusted" logic - I presume for your use case that shouldn't be the default, so...probably all repo finder pulls should just use `OSTREE_REPO_PULL_FLAGS_UNTRUSTED`?

They do already! https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-repo-pull.c#L5367

> Anyways...no opposition to supporting multiple repos, but maybe make the happy path nicer and also support `$usb_stick/.ostree/repo`?

What’s the reasoning for `$usb_stick/.ostree/repo`? Are you thinking of it as an alternative to `$usb_stick/ostree/repo` which is a bit more hidden.

> But in this USB case you're talking about, would there be some logic on the mounting computer to replicate the chain of mounts for live USB? At least in Fedora it's quite interesting.

How do you mean? Are you talking about the pivot root?

**Comment by pwithnall** [comment]
#1179 has been merged, but I’d like to round up with some answers to the questions above before this issue can be closed.

**Comment by cgwalters** [comment]
When I mentioned `$usb_stick/.ostree/repo` I had not internalized that we were looking in `$usb_stick/ostree/repo` - as far as I'm concerned that and with the flatpak change we landed this sounds fine to me.

**Comment by pwithnall** [comment]
Thanks. Closing then!
